File: source/hc_golang_sandbox/ast.go
package main


import (
    . "candlelib/hc_completer"
)


type QueryBody struct {
    tok Token
    container &ContainerClause
filter interface{}
sort interface{}
}
func newQueryBody( tok *Token, _container &ContainerClause, _filter interface{}, _sort interface{}) *QueryBody {
    return &QueryBody{
        tok: *tok,
        container : _container,
        filter : _filter,
        sort : _sort,
    }
}






func (node *QueryBody) Iterate( yield HCNodeIterator, parent HCNode, i int, j int,) {
    
    yield(node, parent, i, j);
        
    
    
}

func (node *QueryBody) Replace(child HCNode, i int, j int){

/*     match i{
    0 => {
                    
        if let Some(old) = self.replace_container(child){ 
                return old;
            }else{
                return ASTNode::NONE;
            }
                }
        _ => {}
    }; */
}


func (node *QueryBody) String() string{
    return node.tok.String()
}

func (node *QueryBody) F64() float64{ return 0 }
func (node *QueryBody) F32() float32{ return 0 }
func (node *QueryBody) I64() int64{ return 0 }
func (node *QueryBody) I32() int32{ return 0 }
func (node *QueryBody) I16() int16{ return 0 }
func (node *QueryBody) I8() int8{ return 0 }

func (node *QueryBody) BOOL() bool{ return false }


func (node *QueryBody) Token() *Token{
    return &node.tok
}
func (node *QueryBody) GetType() uint32{
    return 2
}
func (node *QueryBody) Type() uint32{
    return 2
}



type ContainerClause struct {
    tok Token
    path [] &undefined
container HCNode
}
func newContainerClause( tok *Token, _path [] &undefined, _container HCNode) *ContainerClause {
    return &ContainerClause{
        tok: *tok,
        path : _path,
        container : _container,
    }
}







func (node *ContainerClause) Iterate( yield HCNodeIterator, parent HCNode, i int, j int,) {
    
    yield(node, parent, i, j);
        
    
    
}

func (node *ContainerClause) Replace(child HCNode, i int, j int){

/*     match i{
    0 => {
                    
        if let Some(old) = self.replace_path(child, j){ 
            return old;
        }else{
            return ASTNode::NONE;
        }
                }
    1 => {
                    
        if let Some(old) = self.replace_container(child){ 
                return old;
            }else{
                return ASTNode::NONE;
            }
                }
        _ => {}
    }; */
}


func (node *ContainerClause) String() string{
    return node.tok.String()
}

func (node *ContainerClause) F64() float64{ return 0 }
func (node *ContainerClause) F32() float32{ return 0 }
func (node *ContainerClause) I64() int64{ return 0 }
func (node *ContainerClause) I32() int32{ return 0 }
func (node *ContainerClause) I16() int16{ return 0 }
func (node *ContainerClause) I8() int8{ return 0 }

func (node *ContainerClause) BOOL() bool{ return false }


func (node *ContainerClause) Token() *Token{
    return &node.tok
}
func (node *ContainerClause) GetType() uint32{
    return 4
}
func (node *ContainerClause) Type() uint32{
    return 4
}



type IDENTIFIERS struct {
    tok Token
    ids Vec<Token>
}
func newIDENTIFIERS( tok *Token, _ids Vec<Token>) *IDENTIFIERS {
    return &IDENTIFIERS{
        tok: *tok,
        ids : _ids,
    }
}






func (node *IDENTIFIERS) Iterate( yield HCNodeIterator, parent HCNode, i int, j int,) {
    
    yield(node, parent, i, j);
        
    
    
}

func (node *IDENTIFIERS) Replace(child HCNode, i int, j int){

/*     match i{
    
        _ => {}
    }; */
}


func (node *IDENTIFIERS) String() string{
    return node.tok.String()
}

func (node *IDENTIFIERS) F64() float64{ return 0 }
func (node *IDENTIFIERS) F32() float32{ return 0 }
func (node *IDENTIFIERS) I64() int64{ return 0 }
func (node *IDENTIFIERS) I32() int32{ return 0 }
func (node *IDENTIFIERS) I16() int16{ return 0 }
func (node *IDENTIFIERS) I8() int8{ return 0 }

func (node *IDENTIFIERS) BOOL() bool{ return false }


func (node *IDENTIFIERS) Token() *Token{
    return &node.tok
}
func (node *IDENTIFIERS) GetType() uint32{
    return 6
}
func (node *IDENTIFIERS) Type() uint32{
    return 6
}



func _FN0_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newQueryBody(
        tok,
        (v0).(*ContainerClause),
        nil,
        nil,
    ) 


    return append(args, ref_0)
    
}
func _FN1_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v2 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newContainerClause(
        tok,
        v1,
        (v2).(*IDENTIFIERS),
    ) 


    return append(args, ref_0)
    
}
func _FN2_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newContainerClause(
        tok,
        v0,
        (v1).(*IDENTIFIERS),
    ) 


    return append(args, ref_0)
    
}
func _FN3_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newContainerClause(
        tok,
        nil,
        (v1).(*IDENTIFIERS),
    ) 


    return append(args, ref_0)
    
}
func _FN4_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newContainerClause(
        tok,
        v1,
        nil,
    ) 


    return append(args, ref_0)
    
}
func _FN5_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newContainerClause(
        tok,
        nil,
        (v0).(*IDENTIFIERS),
    ) 


    return append(args, ref_0)
    
}
func _FN6_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newContainerClause(
        tok,
        v0,
        nil,
    ) 


    return append(args, ref_0)
    
}
func _FN7_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := newIDENTIFIERS(
        tok,
        v0,
    ) 


    return append(args, ref_0)
    
}
func _FN8_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    

    return append(args, v0)
    
}
func _FN9_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v2 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    

    val := (v1).String()
    return append(args, HCGObjString{Val:val})
    
}
func _FN10_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := Vec<ASTNode>{ (v0).(Box<IDENTIFIERS>)}

    return append(args, ref_0)
    
}
func _FN11_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := v0.(*[]interface{})
ref_1 := v1
ref_0.push(ref_1);

    return append(args, v0)
    
}
func _FN12_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
ref_0 := &HCGObjStringArray{Val: []string { &(v0).String()}}

    return append(args, ref_0)
    
}
func _FN13_ (args []HCObj, tok *Token) []HCObj { 
    i := len(args)-1
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
    
v0.(*HCGObjStringArray).Append(&(v1).String())

    return append(args, v0)
    
}
func _FN14_ (args []HCObj, tok *Token) []HCObj { return args }
func _FN15_ (args []HCObj, tok *Token) []HCObj {  i := len(args)-1
v1 := args[i]
args = append(args[:i], args[i+1:]...)
i--
v0 := args[i]
args = append(args[:i], args[i+1:]...)
i--
 return append(args, v1); }



var FunctionMaps = []ReduceFunction  {
    _FN14_,
    _FN0_,
    _FN1_,
    _FN2_,
    _FN3_,
    _FN4_,
    _FN5_,
    _FN6_,
    _FN8_,
    _FN7_,
    _FN14_,
    _FN14_,
    _FN9_,
    _FN9_,
    _FN15_,
    _FN14_,
    _FN14_,
    _FN14_,
    _FN14_,
    _FN14_,
    _FN15_,
    _FN14_,
    _FN14_,
    _FN10_,
    _FN11_,
    _FN12_,
    _FN13_,
    _FN12_,
    _FN13_,
}

File: source/hc_rust/sandbox/source/ast.rs


use std::cell::UnsafeCell;

use candlelib_hydrocarbon::ast::{HCObj, HCObjTrait, ReduceFunction};

use candlelib_hydrocarbon::Token;

type RF = ReduceFunction<ASTNode>;

type HCO = HCObj<ASTNode>;


#[derive(Debug, Clone)]
pub enum ASTNode {NONE,QueryBody(Box<QueryBody>),
ContainerClause(Box<ContainerClause>),
IDENTIFIERS(Box<IDENTIFIERS>) 
}
    
impl HCObjTrait for ASTNode {
    fn String(&self) -> String {
        use ASTNode::*;
        match self {
            
            QueryBody(bx) => bx.tok.String(),

            ContainerClause(bx) => bx.tok.String(),

            IDENTIFIERS(bx) => bx.tok.String(),
            _ => String::from(""),
        }
    }
}



#[derive(Debug)]
pub enum NodeIteration<'a> {
    NONE,
    STOP,
    CONTINUE,
    REPLACE(ASTNode),
    QueryBody(&'a mut QueryBody),
ContainerClause(&'a mut ContainerClause),
IDENTIFIERS(&'a mut IDENTIFIERS)
}

impl<'a> NodeIteration<'a> {
    pub fn name(&self) -> &str {
        use NodeIteration::*;
        match self {
            STOP => "stop",
            
                QueryBody(_0) => {
                    "node-QueryBody"
                },
                ContainerClause(_0) => {
                    "node-ContainerClause"
                },
                IDENTIFIERS(_0) => {
                    "node-IDENTIFIERS"
                }
            REPLACE(node) => "replace",
            _ => "unknown",
        }
    }
}

pub trait ASTNodeTraits<'a>
where
    Self: Sized,
{
    fn iterate(
        self: &'a mut Box<Self>,
        _yield: &'a mut impl FnMut(&mut NodeIteration<'a>, &mut NodeIteration<'a>) -> NodeIteration<'a>,
    ) {
        let mut closure = |a: &mut NodeIteration<'a>, b: &mut NodeIteration<'a>, ty:u32, c: i32, d: i32| {
            use NodeIteration::*;
            match _yield(a, b) {
                STOP => false,
                REPLACE(node) => match b {
                    
                    QueryBody(par) => {
                        par.Replace(node, c, d);
                        true
                    },
                    ContainerClause(par) => {
                        par.Replace(node, c, d);
                        true
                    },
                    IDENTIFIERS(par) => {
                        par.Replace(node, c, d);
                        true
                    }
                    _ => true,
                },
                _ => true,
            }
        };

        self.Iterate(&mut closure, &mut NodeIteration::NONE, 0, 0)
    }
    fn Replace(&mut self, node: ASTNode, i: i32, j: i32) -> ASTNode {
        ASTNode::NONE
    }
    fn Iterate(
        self: &'a mut Box<Self>,
        _yield: &mut impl FnMut(&mut NodeIteration<'a>, &mut NodeIteration<'a>,u32, i32, i32) -> bool,
        parent: &mut NodeIteration<'a>,
        i: i32,
        j: i32,
    );
    fn Token(&self) -> Token;
    fn Type() -> u32;
    fn GetType(&self) -> u32;
}




#[derive(Debug, Clone)]
pub struct QueryBody {
    pub tok:Token,
    pub container:Box<ContainerClause>,
pub filter:HCO,
pub sort:HCO
}

impl QueryBody {
fn new( tok: Token, _container:Box<ContainerClause>, _filter:HCO, _sort:HCO) -> Box<Self> {
    Box::new(QueryBody{
        tok: tok,
        container : _container,
        filter : _filter,
        sort : _sort,
    })
}


fn  replace_container(&mut self, child: ASTNode,) -> Option<ASTNode> {
    
    if let ASTNode::ContainerClause(child) = child {
        return Some(ASTNode::ContainerClause(std::mem::replace(&mut self.container, child)))
    }else {
        return None
    }
    
}
}



impl<'a> ASTNodeTraits<'a> for QueryBody
where
    Self: Sized,
{

fn Iterate(
    self: &'a mut Box<Self>,
    _yield: &mut impl FnMut(&mut NodeIteration<'a>, &mut NodeIteration<'a>,u32, i32, i32) -> bool,
    parent: &mut NodeIteration<'a>,
    i: i32,
    j: i32,
) {
    let node = UnsafeCell::from(self);

    {
        let mut_me = unsafe { (*node.get()).as_mut() };

        if !_yield(&mut NodeIteration::QueryBody(mut_me), parent, 2, i, j) { return };
    }
        
    
        {
            let mut_me_b = unsafe { (*node.get()).as_mut() };
            (unsafe { (*node.get()).as_mut() }.container).Iterate( _yield, &mut NodeIteration::QueryBody(mut_me_b), 0, 0);
            
        }
    
}

fn Replace(&mut self, child: ASTNode, i: i32, j: i32) -> ASTNode{

    match i{
    0 => {
                    
        if let Some(old) = self.replace_container(child){ 
                return old;
            }else{
                return ASTNode::NONE;
            }
                }
        _ => {}
    };

    ASTNode::NONE
}


fn Token(&self) -> Token{
    return self.tok;
}

fn Type()-> u32{
    return 2;
}

fn GetType(&self) -> u32 {
    return 2;
}
}




#[derive(Debug, Clone)]
pub struct ContainerClause {
    pub tok:Token,
    pub path:Vec<Box<IDENTIFIERS>>,
pub container:ASTNode
}

impl ContainerClause {
fn new( tok: Token, _path:Vec<Box<IDENTIFIERS>>, _container:ASTNode) -> Box<Self> {
    Box::new(ContainerClause{
        tok: tok,
        path : _path,
        container : _container,
    })
}


fn  replace_path(&mut self, child: ASTNode,index: i32,) -> Option<ASTNode> {
    
    match &child {
        ASTNode::IDENTIFIERS(_) => {
            if index as usize >= self.path.len() {
                self.path.push(child);
                None
            }else {
                self.path.push(child);
                let node = self.path.swap_remove(index as usize);
                Some(node)
            }
        }
        ASTNode::NONE => {
            if (index as usize)< self.path.len() {
                let node = self.path.remove(index as usize);
                Some(node)
            }else {
                None
            }
        }
        _ => None
    }
}

fn  replace_container(&mut self, child: ASTNode,) -> Option<ASTNode> {
    
    match &child {
        ASTNode::NONE => {
            let old = std::mem::replace(&mut self.container, ASTNode::NONE);
            return Some(old);
        }
        
        ASTNode::IDENTIFIERS(_) => { 
            return Some(std::mem::replace(&mut self.container, child));
        },

        ASTNode::IDENTIFIERS(_) => { 
            return Some(std::mem::replace(&mut self.container, child));
        },

        ASTNode::IDENTIFIERS(_) => { 
            return Some(std::mem::replace(&mut self.container, child));
        }
        _ => None
        
    }
}
}



impl<'a> ASTNodeTraits<'a> for ContainerClause
where
    Self: Sized,
{

fn Iterate(
    self: &'a mut Box<Self>,
    _yield: &mut impl FnMut(&mut NodeIteration<'a>, &mut NodeIteration<'a>,u32, i32, i32) -> bool,
    parent: &mut NodeIteration<'a>,
    i: i32,
    j: i32,
) {
    let node = UnsafeCell::from(self);

    {
        let mut_me = unsafe { (*node.get()).as_mut() };

        if !_yield(&mut NodeIteration::ContainerClause(mut_me), parent, 4, i, j) { return };
    }
        
    
        {

            let mut_me_a = unsafe { (*node.get()).as_mut() };
            for j in 0..mut_me_a.path.len() {
                let mut_me_b = unsafe { (*node.get()).as_mut() };
                let child = &mut mut_me_b.path[j];

                
                if let ASTNode::IDENTIFIERS(child) = child {
                    let mut_me = unsafe { (*node.get()).as_mut() };
                    child.Iterate( _yield, &mut NodeIteration::ContainerClause(mut_me), 0, j as i32);
                }
            }
        }
    
        {
            if let Some(child) = &mut (unsafe { (*node.get()).as_mut() }.container) {
                let mut_me_b = unsafe { (*node.get()).as_mut() };
                child.Iterate(_yield, &mut NodeIteration::ContainerClause(mut_me_b), 1, 0);
            }
        }
    
}

fn Replace(&mut self, child: ASTNode, i: i32, j: i32) -> ASTNode{

    match i{
    0 => {
                    
        if let Some(old) = self.replace_path(child, j){ 
            return old;
        }else{
            return ASTNode::NONE;
        }
                }
    1 => {
                    
        if let Some(old) = self.replace_container(child){ 
                return old;
            }else{
                return ASTNode::NONE;
            }
                }
        _ => {}
    };

    ASTNode::NONE
}


fn Token(&self) -> Token{
    return self.tok;
}

fn Type()-> u32{
    return 4;
}

fn GetType(&self) -> u32 {
    return 4;
}
}




#[derive(Debug, Clone)]
pub struct IDENTIFIERS {
    pub tok:Token,
    pub ids:Vec<Token>
}

impl IDENTIFIERS {
fn new( tok: Token, _ids:Vec<Token>) -> Box<Self> {
    Box::new(IDENTIFIERS{
        tok: tok,
        ids : _ids,
    })
}


}



impl<'a> ASTNodeTraits<'a> for IDENTIFIERS
where
    Self: Sized,
{

fn Iterate(
    self: &'a mut Box<Self>,
    _yield: &mut impl FnMut(&mut NodeIteration<'a>, &mut NodeIteration<'a>,u32, i32, i32) -> bool,
    parent: &mut NodeIteration<'a>,
    i: i32,
    j: i32,
) {
    let node = UnsafeCell::from(self);

    {
        let mut_me = unsafe { (*node.get()).as_mut() };

        if !_yield(&mut NodeIteration::IDENTIFIERS(mut_me), parent, 6, i, j) { return };
    }
        
    
    
}

fn Replace(&mut self, child: ASTNode, i: i32, j: i32) -> ASTNode{

    match i{
    
        _ => {}
    };

    ASTNode::NONE
}


fn Token(&self) -> Token{
    return self.tok;
}

fn Type()-> u32{
    return 6;
}

fn GetType(&self) -> u32 {
    return 6;
}
}



fn _fn0 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v0 = args.remove(i-0);
                                
 if let HCO::NODE(r_0) = v0 { 
 if let ASTNode::ContainerClause(r_1) = r_0 { 
 let mut ref_0 = HCO::NODE(ASTNode::QueryBody(QueryBody::new(
        tok,
        r_1,
        Vec::new(),
        Vec::new(),
    ) 
));;

                                args.push(ref_0);  } }
                            }
fn _fn1 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v2 = args.remove(i-0);
let mut v1 = args.remove(i-1);
let mut v0 = args.remove(i-2);
                                
 if let HCO::NODES(r_0) = v1 { 
 if let HCO::NODE(r_1) = v2 { 
 let mut ref_0 = HCO::NODE(ASTNode::ContainerClause(ContainerClause::new(
        tok,
        r_0,
        r_1,
    ) 
));;

                                args.push(ref_0);  } }
                            }
fn _fn2 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v1 = args.remove(i-0);
let mut v0 = args.remove(i-1);
                                
 if let HCO::NODES(r_0) = v0 { 
 if let HCO::NODE(r_1) = v1 { 
 let mut ref_0 = HCO::NODE(ASTNode::ContainerClause(ContainerClause::new(
        tok,
        r_0,
        r_1,
    ) 
));;

                                args.push(ref_0);  } }
                            }
fn _fn3 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v1 = args.remove(i-0);
let mut v0 = args.remove(i-1);
                                
 if let HCO::NODE(r_0) = v1 { 
 let mut ref_0 = HCO::NODE(ASTNode::ContainerClause(ContainerClause::new(
        tok,
        Vec::new(),
        r_0,
    ) 
));;

                                args.push(ref_0);  }
                            }
fn _fn4 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v1 = args.remove(i-0);
let mut v0 = args.remove(i-1);
                                
 if let HCO::NODES(r_0) = v1 { 
 let mut ref_0 = HCO::NODE(ASTNode::ContainerClause(ContainerClause::new(
        tok,
        r_0,
        Vec::new(),
    ) 
));;

                                args.push(ref_0);  }
                            }
fn _fn5 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v0 = args.remove(i-0);
                                
 if let HCO::NODE(r_0) = v0 { 
 let mut ref_0 = HCO::NODE(ASTNode::ContainerClause(ContainerClause::new(
        tok,
        Vec::new(),
        r_0,
    ) 
));;

                                args.push(ref_0);  }
                            }
fn _fn6 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v0 = args.remove(i-0);
                                
 if let HCO::NODES(r_0) = v0 { 
 let mut ref_0 = HCO::NODE(ASTNode::ContainerClause(ContainerClause::new(
        tok,
        r_0,
        Vec::new(),
    ) 
));;

                                args.push(ref_0);  }
                            }
fn _fn7 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v0 = args.remove(i-0);
                                
 let mut ref_0 = HCO::NODE(ASTNode::IDENTIFIERS(IDENTIFIERS::new(
        tok,
        v0,
    ) 
));;

                                args.push(ref_0); 
                            }
fn _fn8 (args:&mut Vec<HCO>, tok: Token){ 
                        let mut i = args.len()-1;
let mut v1 = args.remove(i-0);
let mut v0 = args.remove(i-1);
                                

                                args.push(v0); 
                            }
fn _fn9 (args:&mut Vec<HCO>, tok: Token){ let mut i = args.len()-1;
let mut v2 = args.remove(i-0);
let mut v1 = args.remove(i-1);
let mut v0 = args.remove(i-2); 
 return HCO::STRING((v1).String())}
fn _fn10 (args:&mut Vec<HCO>, tok: Token){ 
                                let mut i = args.len()-1;
let mut v0 = args.remove(i-0);
                                
 let mut ref_0:Vec<ASTNode> = Vec::new();
 if let HCO::NODE(r_0) = v0 { 
ref_0.push(r_0);
 }
                                args.push(HCO::NODES(ref_0)); } 
fn _fn11 (args:&mut Vec<HCO>, tok: Token){ 
                                let mut i = args.len()-1;
let mut v1 = args.remove(i-0);
let mut v0 = args.remove(i-1);
                                
 if let HCO::NODES(r_0) = &mut v0 { 
 if let HCO::NODE(r_1) = v1 { 
r_0.push(r_1);
  } }
                                args.push(v0) } 
fn _fn12 (args:&mut Vec<HCO>, tok: Token){
                            let mut i = args.len()-1;
let mut v0 = args.remove(i-0); 
                                
 let mut ref_0 = &HCGObjStringArray{Val: [String] { &(v0).String()}};

                                args.push(ref_0); 
                            }
fn _fn13 (args:&mut Vec<HCO>, tok: Token){
                            let mut i = args.len()-1;
let mut v1 = args.remove(i-0);
let mut v0 = args.remove(i-1); 
                                
v0.(*HCGObjStringArray).Append(&(v1).String())

                                args.push(v0); 
                            }
fn _fn14 (args:&mut Vec<HCO>, tok: Token){}
fn _fn15 (args:&mut Vec<HCO>, tok: Token){  let mut i = args.len()-1;
let mut v1 = args.remove(i-0);
let mut v0 = args.remove(i-1);
 args.push(v1); }



pub const FunctionMaps:[RF; 29]= [
    _fn14,_fn0,_fn1,_fn2,_fn3,_fn4,_fn5,_fn6,_fn8,_fn7,_fn14,_fn14,_fn9,_fn9,_fn15,_fn14,_fn14,_fn14,_fn14,_fn14,_fn15,_fn14,_fn14,_fn10,_fn11,_fn12,_fn13,_fn12,_fn13,
];
